### PR TITLE
Move URL matching into its own class

### PR DESCRIPTION
**Pull request by jrmsklar**
This moves the URL matching logic out of `URLNavigator` and into its own class `URLMatcher`.

The motivation behind this was to decouple the matching and navigating logic so that we could use the URL matcher by itself. Our API layer requires the ability to match URLs.

A couple other additions that this provides:
- URL value matching against [`UUID`](https://en.wikipedia.org/wiki/Universally_unique_identifier) and `string` types, and any other custom type (see https://github.com/devxoul/URLNavigator/commit/90b08b26b6a7d69a2450dc74a32dd400ed7f6789)
- Query item parsing when matching URLs (see https://github.com/devxoul/URLNavigator/commit/0681473e1875741fcc60c393daff82d4eac01240)
- Along with the bullet point above comes a typed return value (`URLMatchComponents`) from `matchURL(_:scheme:from:)` (see https://github.com/devxoul/URLNavigator/commit/26fe9b3091586c6131681bc39f06087a45b69e50)

Thanks to @mpiccinato for some help with this!

### CODE REVIEW

File: Sources/URLMatcher.swift [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+    ///     matcher.addURLValueMatcherHandler("SSN") { (ssnString) -> AnyObject? in
+    ///         return SSN(string: ssnString)
+    ///     }
+    ///
+    /// The value type that this would match against is "ssn" (i.e. Social Security Number), and the
+    /// handler to be used for that type returns a newly created `SSN` object from the ssn string.
+    ///
+    /// - Parameter valueType: The value type (string) to match against.
+    /// - Parameter handler: The handler to use when matching against that value type.
+    public func addURLValueMatcherHandler(valueType: String, handler: URLValueMatcherHandler) {
```
Comment (devxoul) [inline Sources/URLMatcher.swift]: I really love this approach :+1:

**Comment by devxoul** [comment]
@jrmsklar, this is really cool. Thanks for your work! I'll add some review on comments. :smile::+1:

**Comment by devxoul** [comment]
This is awesome. Everything is perfect. Thanks @jrmsklar!

**Comment by jrmsklar** [comment]
Thanks @devxoul! Nice work creating and maintaining this repo. It's a great resource and well put together 👍.

**Pull request merged**